Support link arrays on links table at entity editor page

yusufkinatas opened the pull request:
## 🌟 What is the purpose of this PR?

This PR expands the functionality of the **links table** on the **edit entity** page. With these changes, users can link and unlink entities on link lists (links where the value of `maxItems` is greater than `1`). This PR also changes the links table design based on the current status of (ongoing) internal discussions.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/0/1203199644802795) _(internal)_
- [Design](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH-App?node-id=6468%3A253538&t=5AeJC1C6PDLgoPPZ-4) _(internal)_
  - ⚠️ The implemented designs deviate from the originals in a number of places. Follow-up decisions need to be made to adjust either the designs or the code. Follow-up tickets will be created.
- [Slack thread](https://hashintel.slack.com/archives/C022217GAHF/p1670414417826489) _(internal)_ where we discuss some of the changes

## 🔍 What does this change?

- Add optional `opacity` support to `drawCellFadeOutGradient` function
- Design changes on links table
- Make cursor `pointer` on `linkedWith` column
- Add `title` prop to `AddAnotherButton` component
- Add `icon` & `tooltip` props to `ValueChip` component

## ⚠️ Known issues

There are some known issues, most of them being UX related. We need to discuss how to solve those issues efficiently without wasting effort. Because we'll eventually implement two big changes on top of entity editor.

Those big changes are `collab support` and `draft entity state`. These two can potentially change the entity-editor's state management drastically, so it could be clever to avoid spending more effort on fixing UX issues on the links table if those UX issues will be solved by the changes we'll introduce with `collab` & `draft entity state`.

### Issues of links table
Removing / adding links take some time. And we’re not doing actual optimistic updates, because it’s hard to update the entity subgraph manually for links. So instead of optimistic updates, we just `refetch` the entity subgraph on changes. That causes some UX issues. 

### Technical explanation of those UX issues
If users link/unlink entities in quick succession, race conditions happen. Since we didn’t implemented any loading state for the table due to avoid wasting effort. Because eventually table will update immediately without a need for loading state. Eventually we’ll d (A) optimistic update or (B) draft entity state managed on client-side.

We decided to open this PR with it’s current state with @CiaranMn, because it has the actual functionality working, even if with some UX flaws. If users does not try to link/unlink entities in a quick succession, it should be fine.

### Some demos of the UX issues / bugs happen when users link/unlink entities in quick succession

https://user-images.githubusercontent.com/39553853/206283861-8b2eef2b-6651-4b46-aec0-926b4b386e58.mp4



## 🐾 Next steps

- Fix the known issues by adressing them individually, or wait for `collab` & `draft entity state` changes to see how the code is going to change around these issues.

## ❓ How to test this?


1.  Checkout the branch / view the deployment
1.  Open up entity editor for any entity, and use the [create demo entity type button](https://github.com/hashintel/hash/commit/202d6c0103a89e13bd8e75a9951cbb22945b29ee) to create a demo entity type
1.  Confirm that links table works as shown in the demo below.

## 📹 Demo

https://user-images.githubusercontent.com/39553853/206265294-3157c259-3dbb-45a2-b7a6-1108e8db68a4.mov

yusufkinatas commented on packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/entity-selector.tsx:
```diff
@@ -0,0 +1,135 @@
+    void init();
+  }, [aggregateEntities, expectedEntityTypes]);
+
+  const sortedAndFilteredEntities = useMemo(() => {
+    return [...entities]
+      .filter(
+        (entity) =>
+          !entityIdsToFilterOut.includes(entity.metadata.editionId.baseId),
+      )
+      .sort((a, b) =>
+        a.metadata.editionId.baseId.localeCompare(b.metadata.editionId.baseId),
+      );
+  }, [entities, entityIdsToFilterOut]);
```
On this PR, only new part to this code it this filter & sort part. Rest of this file is copy-paste from it's previous location.

nathggns commented:
Reviewing now

nathggns commented:
When you move the mouse up from the delete icon into the blank space, the delete tooltip doesn't hide. This is really unexpected and annoying from a UX perspective if you're trying to reach the delete icon above covered by the tooltip. 
<img width="151" alt="image" src="https://user-images.githubusercontent.com/719814/206454156-4d86f0a5-617c-420b-bc40-c737e24388c9.png">

nathggns commented:
What does this "selected" state on a chip indicate? 

<img width="512" alt="image" src="https://user-images.githubusercontent.com/719814/206454576-e79ea2e1-ab26-40a4-b3ca-3ec437514f2c.png">

yusufkinatas commented:
> What does this "selected" state on a chip indicate?

Without it, users can see delete button on mobile. Because there is no hover on mobile

nathggns commented:
<img width="1041" alt="image" src="https://user-images.githubusercontent.com/719814/206454915-47acb83c-cae0-4d46-b0c3-f923b01d9567.png">

There's a double border on the select entity field and it's also slightly too big on either side, and there's something whacky going on with the curved edges

nathggns commented:
@yusufkinatas Thanks – I struggled to follow what exactly the video was indicating.

nathggns commented:
Ah, I didn't realise it had sound.

yusufkinatas commented on packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/entity-selector.tsx:
```diff
@@ -0,0 +1,135 @@
+          !entityIdsToFilterOut.includes(entity.metadata.editionId.baseId),
+      )
+      .sort((a, b) =>
+        a.metadata.editionId.baseId.localeCompare(b.metadata.editionId.baseId),
```
Oops, it meant to be `editionId.version`, thanks for catching this!

yusufkinatas commented on packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-entity-list-editor.tsx:
```diff
@@ -0,0 +1,163 @@
+  } = cell.data.linkRow;
+
+  const [addingLink, setAddingLink] = useState(!linkAndTargetEntities.length);
+  const [selectedLinkEntityId, setSelectedLinkEntityId] = useState("");
```
I guess making it `null` makes it less error-prone for future right? e.g. we can use `??` operator for null safely.

yusufkinatas commented on packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-entity-list-editor.tsx:
```diff
@@ -0,0 +1,163 @@
+        overflow: "hidden",
+      })}
+    >
+      <Box sx={{ maxHeight: 300, overflowY: "scroll" }}>
```
Definitely!

yusufkinatas commented:
Thanks for the review, addressed all of the comments above with this commit https://github.com/hashintel/hash/pull/1614/commits/04faed66482ff379cde4bbb990d4102206b461a6

nathggns commented:
@yusufkinatas the tooltip is flashing in the top left corner https://www.loom.com/share/b822f7e352884e6ab971cf1f3216b5d5

luisbettencourt commented on packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/linked-entity-list-editor/max-items-reached.tsx:
```diff
@@ -0,0 +1,15 @@
+
+export const MaxItemsReached = ({ limit }: { limit: number }) => {
+  return (
+    <Button
```
Why is this a `Button`, is it meant to be interactible in the future?

yusufkinatas replied:
To achieve the same UI, but in a disabled state.

luisbettencourt commented on packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/links-section/link-table/cells/linked-with-cell/entity-selector.tsx:
```diff
@@ -0,0 +1,137 @@
+        (entity) =>
+          !entityIdsToFilterOut?.includes(entity.metadata.editionId.baseId),
+      )
+      .sort((a, b) =>
+        a.metadata.editionId.version.localeCompare(
+          b.metadata.editionId.version,
+        ),
```
You could probably reuse `sortLinkAndTargetEntities` here
```suggestion
      .sort((a, b) =>
        a.metadata.editionId.version.localeCompare(
          b.metadata.editionId.version,
        ),
```

luisbettencourt commented on packages/hash/frontend/src/pages/[account-slug]/entities/[entity-uuid].page/entity-editor/properties-section/property-table/cells/value-cell/value-cell-editor/array-editor/sortable-row.tsx:
```diff
@@ -111,7 +112,12 @@ export const SortableRow = ({
+        <ValueChip
+          value={value}
+          selected={selected}
+          icon={{ icon: faText }}
+          tooltip="Text"
```
I expect this icon and tooltip text to change according to data type, right? Is this being left out for a future PR?

yusufkinatas replied:
That has it's own ticket on Asana, but thanks!

luisbettencourt reviewed (changes requested):
Awesome work, @yusufkinatas! Editing links already feels great!

Tested it with the demo entity you provided and with comment entities as well, working as expected. Left a couple of small comments.

luisbettencourt reviewed (dismissed):
Good to go for me, but I'd wait for @nathggns approval aswell.

yusufkinatas commented:
@nathggns Tooltip flickering issue should be fixed with this commit https://github.com/hashintel/hash/pull/1614#:~:text=use%20visibility%20css%20prop%20instead%20of%20display
